Avoid decomposing `_unsafe_index` in Inductor

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107882

`_unsafe_index` was previously added to the core ATen decomp table in https://github.com/pytorch/pytorch/pull/106814, but this has performance ramifications for Inductor. Therefore, this diff removes it from the decomposition table used by Inductor.

Differential Revision: [D48649210](https://our.internmc.facebook.com/intern/diff/D48649210/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Commit:
Avoid decomposing `_unsafe_index` in Inductor

`_unsafe_index` was previously added to the core ATen decomp table in https://github.com/pytorch/pytorch/pull/106814, but this has performance ramifications for Inductor. Therefore, this diff removes it from the decomposition table used by Inductor.

Differential Revision: [D48649210](https://our.internmc.facebook.com/intern/diff/D48649210/)

[ghstack-poisoned]

SherlockNoMad commented on torch/_inductor/decomposition.py:
```diff
@@ -56,6 +56,9 @@
 
+# Remove unwanted decompositions included via the core ATen decompositions from
+# the Inductor decomp table.
+decompositions.pop(aten._unsafe_index.Tensor, None)
```
Please create a exclude_decomp_list, and iterate over the list to exclude. 
We have have more then one decomp to exclude.

SherlockNoMad reviewed (approved):
Per comment.

Commit:
Update on "Avoid decomposing `_unsafe_index` in Inductor"


`_unsafe_index` was previously added to the core ATen decomp table in https://github.com/pytorch/pytorch/pull/106814, but this has performance ramifications for Inductor. Therefore, this diff removes it from the decomposition table used by Inductor.

Differential Revision: [D48649210](https://our.internmc.facebook.com/intern/diff/D48649210/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Commit:
Update on "Avoid decomposing `_unsafe_index` in Inductor"


`_unsafe_index` was previously added to the core ATen decomp table in https://github.com/pytorch/pytorch/pull/106814, but this has performance ramifications for Inductor. Therefore, this diff removes it from the decomposition table used by Inductor.

Differential Revision: [D48649210](https://our.internmc.facebook.com/intern/diff/D48649210/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/decomposition.py:
```diff
@@ -56,6 +60,14 @@
+    aten._unsafe_index,
+]
+
+remove_decompositions(decompositions, decomps_to_exclude)
```
Do we not expect exported models to never be run under inductor? Otherwise this seems like a bad solution.

angelayi replied:
We expect some models to be run under inductor, but we want export to be a generic representation. Export allows users to customize their decompositions (ex. running inductor decomps instead of core aten decomps).
